### PR TITLE
Implement a new Async Server approach based on pyzmq to get completions on the Editor

### LINKED ISSUE #3033: Create a Remote Procedure Call helper
This will be used for pylint, pep8, and introspection.
- Create a QProcess with the executable of the current project, adding spyder and the appropriate dependencies (e.g. jedi) to the path. 
- Use ZMQ to pass pickled objects to a remote procedure call.
- The remote end imports the desired modules(s) and performs the remote procedure calls, returning the result over the ZMQ socket pair.

```
RemoteClient(executable, dependencies, imports)
    onReply(request_number, result)
    send(func_name, *args, **kwargs)  # returns a request number
RemoteServer(imports, server_port, client_port)
```

### LINKED ISSUE #3041: Spyder crash with "too many files open" message
## Description

**What steps will reproduce the problem?**

1. Open spyder
2. Live your life
3. Return to spyder: it's gone

**What is the expected output? What do you see instead?**
I've never seen the crash, because each time I was on another desktop when it happened. I had this problem around 4 times in less than a week. This problem appeared last week, considering the rate at which I upgrade spyder on this computer the problem is no more than say 3 weeks old.

Here is the message I get in the console:
```
We got some errors while running testparm "Load smb config files from /etc/samba/smb.conf\nError loading services.\n"
KSambaShare: Could not find smb.conf!
We got some errors while running 'net usershare info'
"Can't load /etc/samba/smb.conf - run testparm to debug it\n"
QProcessPrivate::createPipe: Cannot create pipe 0x3d42978: Too many open files
Unhandled Python exception
fish: 'spyder3' terminated by signal SIGABRT (Abandon)
```
I don't know why it complains about smb, and I don't know if it's related. All the files I open are on the machine itself.

**Please provide any additional information below**

## Version and main components

* Spyder Version: 3.0.0dev 
* Python Version: 3.5.1
* Qt Versions: 5.5.1, PyQt5 5.5.1 on Linux

## Dependencies
```
pyflakes >=0.6.0:  1.0.0 (OK)
pep8 >=0.6      :  1.7.0 (OK)
pygments >=1.6  :  2.1.1 (OK)
qtconsole >=4.0 :  4.1.1 (OK)
nbconvert >=4.0 :  4.1.0 (OK)
pandas >=0.13.1 :  0.17.1 (OK)
numpy >=1.7     :  1.10.4 (OK)
sphinx >=0.6.6  :  1.3.5 (OK)
psutil >=0.3    :  4.0.0 (OK)
rope >=0.9.4    :  0.9.4-1 (OK)
jedi >=0.8.1    :  0.9.0 (OK)
matplotlib >=1.0:  1.5.1 (OK)
sympy >=0.7.3   :  None (NOK)
pylint >=0.25   :  1.5.4 (OK)

```

### PR DESCRIPTION
**Pull request by blink1073**
Implements streamlined async server for introspection, fixes #3033, fixes #3041.

Uses a single port for communication that is managed by zmq.
Uses socket notification rather than a thread in the client.
Adds a heartbeat to keep the server alive.
The main classes will be moved elsewhere and used for code analysis background tasks after beta3.

### CODE REVIEW

**Comment by ccordoba12** [comment]
@blink1073, thanks a lot for working on this. I'd like to know what you mean by

> partially address #3033 

What's left to be done to solve that issue?

----

Also, please add `pyzmq` to our setup.py and conda recipe and README file, to reflect that now it is a first class dependency (instead of being inherited through qtconsole).

**Comment by blink1073** [comment]
@ccordoba12, I'll close the other issue with this and make a new one specific to analysis tasks for clarity.

**Comment by ccordoba12** [comment]
Ok, nice! Please remember that this PR also fixes #3041 :-)

**Comment by ccordoba12** [comment]
And sorry for the confusion, I thought #3033 == #3041, so I was a bit worried that something was missing to fix #3041 (which is a serious issue).

**Comment by blink1073** [comment]
@ccordoba12, done on all counts.

File: spyderlib/utils/introspection/plugin_client.py [hunk 1]
```diff
@@ -4,165 +4,203 @@
...
+        self.closing = False
+        self.context = zmq.Context()
+        QApplication.instance().aboutToQuit.connect(self.close)
 
-    def start(self):
-        """Start a new connection with the plugin.
+        # Set up the heartbeat timer.
+        self.timer = QTimer(self)
+        self.timer.timeout.connect(self._heartbeat)
+        self.timer.start(HEATBEAT)
```
Comment (goanpeca) [inline spyderlib/utils/introspection/plugin_client.py]: heatbeat? is that a new ryrthm?

Comment (blink1073) [inline spyderlib/utils/introspection/plugin_client.py]: https://www.youtube.com/watch?v=TpqAUtzWDuw

**Comment by blink1073** [comment]
I verified this works on Windows 7.

File: README.md [hunk 2]
```diff
@@ -139,17 +139,18 @@ a Python version greater than 2.7 (Python 3.2 is not supported anymore).
 * **Python** 2.7 or 3.3+
 * **PyQt5** 5.2+ or **PyQt4** 4.6+: PyQt5 is recommended.
 * **qtconsole**: Enhanced Python interpreter.
-* **Rope** and **Jedi**: Editor code completion, calltips
+* **Rope** and **Jedi**: Editor code completion, calltips.
```
Comment (ccordoba12) [inline README.md]: There's no need to put a dot at the end here. Please notice that the text continues in the next line :-)

File: spyderlib/utils/introspection/plugin_server.py [hunk 3]
```diff
@@ -4,101 +4,95 @@
...
-        """Handle requests"""
+        """Handle requests from the client.
+        """
         while 1:
-            # Get most recent request
-            request = None
+            # Poll for events, handling a timeout.
+            events = self.socket.poll(TIMEOUT)
+            if events == 0:
+                print('Timed out')
```
Comment (ccordoba12) [inline spyderlib/utils/introspection/plugin_server.py]: Please remove this print (or make it a `debug_print`)

File: spyderlib/utils/introspection/plugin_server.py [hunk 4]
```diff
@@ -4,101 +4,95 @@
...
+                return
+            # Drain all exising requests, handling quit and heartbeat.
+            requests = []
             while 1:
-                try:
-                    request = self.queue.get(True, 0.01)
-                except Queue.Empty:
+                request = self.socket.recv_pyobj()
+                if request['func_name'] == 'server_quit':
+                    print('Quitting')
```
Comment (ccordoba12) [inline spyderlib/utils/introspection/plugin_server.py]: Also remove this `print`

Comment (blink1073) [inline spyderlib/utils/introspection/plugin_server.py]: This is being run in a separate process, with a separate stdout..  We want the print so we can capture it from the client in the event of an error.

**Comment by ccordoba12** [comment]
This is working very nice, thanks @blink1073!

I'm just getting this error in the internal console after a completion request is performed and succeeds:

```python-traceback
Traceback (most recent call last):
  File "/home/carlos/miniconda/envs/test-async/lib/python2.7/site-packages/spyderlib/utils/introspection/manager.py", line 125, in handle_response
    self._finalize(response)
  File "/home/carlos/miniconda/envs/test-async/lib/python2.7/site-packages/spyderlib/utils/introspection/manager.py", line 137, in _finalize
    % (self.info.name, response['name'],
AttributeError: 'NoneType' object has no attribute 'name'
```

This happens on Linux, with PyQ4 and PyQt5, and Python 2.7

File: README.md (hunk 2 above)
Comment (blink1073) [inline README.md]: Fixed.

File: spyderlib/utils/introspection/plugin_server.py (hunk 4 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/plugin_server.py]: Ok, I assume the above `print` is the same ;-)

Comment (blink1073) [inline spyderlib/utils/introspection/plugin_server.py]: Yes.

**Comment by blink1073** [comment]
Does the latest push address the error?

File: spyderlib/utils/introspection/plugin_server.py (hunk 4 above)
Comment (ccordoba12) [inline spyderlib/utils/introspection/plugin_server.py]: :+1:

**Comment by ccordoba12** [comment]
Yes, it does. But now there's a problem when I start Spyder with `SPYDER_DEBUG=3`: it frozes after the main window is created. Before your last change, it didn't happen that.

Could you verify it in your side?

**Comment by blink1073** [comment]
Couldn't verify, but tried another fix, care to try again?

**Comment by ccordoba12** [comment]
Thanks, your last commit fixes the freeze I was seeing. So, as soon as Travis and AppVeyor are green, I'll merge ;-)

**Pull request merged**